rfcs/20221020-tensor-indexed-updates.md

```diff
@@ -0,0 +1,468 @@
+```python
+# Proposed tf.Variable in-place slice assignment:
+v[:, :, 2] = 0
+```
+This would modify the `tf.Variable` in-place.  For consistency with the proposed
+`Tensor` API, we would also allow
+```python
+# Proposed tf.Variable copy-and-update assignment:
+v = v.at[:, :, 2].set(0)
+```
+Which would create a modified copy of the original variable.
+
+
+## User Benefit
+
+Users will finally be able to "update" tensors by index using a friendly
+NumPy-like slice syntax, while also keeping `Tensor`s immutable.  This is part
+of a larger direction to introduce more Numpy-like behaviors in TensorFlow to
+improve compatibility and the user experience.
+
+## Design Proposal
+
+We are proposing the addition of a new python API only - no new ops need to be
+added, though some may be updated to support broadcasting to take full
+advantage of efficient updates.  For consistency with NumPy and JAX, this API
+should handle the following types of indexing:
+
+- Basic indexing
+  - single element: `x[1, 2, 2]`
+  - slicing/striding: `x[:, 1:2:4, 2]`
+- Advanced indexing
+  - integer tensor indexing: `x[ [[1, 2], [3, 4], [5, 6]] ]`
+  - boolean array indexing: `x[x < 0]`
+
+We are _not_ currently targetting NumPy's [_combined_ basic and
+advanced indexing](https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing), since this is not currently
+supported by our `Tensor.__getitem__` method, and this has not been identified
+as a common use-case.
+
+We will use the JAX `.at[]` syntax to access the elements to update.
+Specifically, we will:
+- Add a `.at` helper property to `tf.Tensor` and related
+  types to enable this indexing behavior.
+- For `tf.Variable`, overload `tf.Variable.__setitem__` to allow slice
+  assignment directly.
+
+Under the hood, these will call into existing tensorflow methods and kernels to
+perform the copy-and-update operation.  See [Detailed Design](#detailed-design)
+for more details about the implementation.
+
+Along with `set(values)`, for consistency with JAX we will also provide:
+- `add(values)`
+- `multiply(values)`
+- `divide(values)`
+- `power(values)`
+- `min(values)`
+- `max(values)`
+- `apply(ufunc)`
+- `get()`
+
+The initial implementation for these will simply use a get-modify-set pattern.
+If use-cases are discovered where we need performant implementations, new ops
+and kernels can be added in an ad-hoc fashion.  Note that performance can also
+be achieved by taking advantage of XLA and jit-compilation via
+`tf.function(jit_compile=true)` to fuse the operations internally.
+
+### Alternatives Considered
+
+#### Copy-And-Update Function
+
+The main alternative considered was to introduce a new function:
+```python
+tf.copy_and_update(input_tensor, index_expression, updates)
+```
+The main difficulty is in constructing the `index_expression`.  NumPy _does_
+already have a convenience routine for this: [`np.index_exp`](https://numpy.org/doc/stable/reference/generated/numpy.s_.html),
```

Comment:
"a convenient routine"

Reply:
https://en.wikipedia.org/wiki/Convenience_function

I'll change to "convenience function"